File: fenwick-tree/cplusplus/1112.cpp
#include <bits/stdc++.h>

using namespace std;

/////////////////////////////////////////////////
typedef long long Int;
typedef pair<int, int> pii;
typedef vector<int> vi;
/////////////////////////////////////////////////
#define REP(i, n) \
	for (int i = 0; i < (int)n; ++i)
#define FOR(i, n) \
	for (int i = 1; i <= (int)n; ++i)
#define ITR(c) __typeof((c).begin())
#define foreach(i, c) \
	for (ITR(c) i = (c).begin(); i != (c).end(); ++i)
#define ALL(c) \
	(c).begin(), (c).end()
#define DB(x) \
	cout << #x << " = " << x << endl

#define endl '\n'
#define F first
#define S second
#define pb push_back
#define mp make_pair
#define LEFT(n) ((n << 1) + 1)
#define RIGHT(n) ((n << 1) + 2)
#define BIT(n) (1 << n)
#define ONES(n) __builtin_popcount(n)
#define rightZero(n) __builtin_ctz(n); // trailing zeros
#define leftZero(n) __builtin_clz(n);	// leading zeros

/////////////////////////////////////////////////
const double EPS = 1e-15;
const int oo = (1 << 30);
const double PI = M_PI;
const int MOD = 1000000000 + 7;
/////////////////////////////////////////////////

const int
		MaxN = 100001;

struct abi
{

	vector<int> ft;
	abi(int len) : ft(len + 1, 0) {}

	int lowbit(int x) { return x & -x; }

	void update(int pos, int val)
	{
		for (; pos < (int)ft.size(); pos += lowbit(pos))
			ft[pos] += val;
	}

	int query(int pos)
	{
		int sum = 0;
		for (; pos > 0; pos -= lowbit(pos))
			sum += ft[pos];
		return sum;
	}

	int query(int l, int r)
	{
		l = (l > 0) ? l - 1 : 0;
		return query(r) - query(l);
	}
};

int main()
{
	//ios_base::sync_with_stdio(false);
	//cin.tie(0);

	int t;
	scanf("%d", &t);
	for (int cs = 1; cs <= t; ++cs)
	{
		int n, q;
		scanf("%d %d", &n, &q);
		abi f(n);

		for (int i = 0; i < n; ++i)
		{
			int ni;
			scanf("%d", &ni);
			f.update(i + 1, ni);
		}

		printf("Case %d:\n", cs);
		for (int i = 0; i < q; ++i)
		{
			int op;
			scanf("%d", &op);

			if (op == 1)
			{
				int pos;
				scanf("%d", &pos);
				int sack = f.query(pos + 1, pos + 1);

				f.update(pos + 1, -sack);
				printf("%d\n", sack);
			}
			else if (op == 2)
			{
				int pos, val;
				scanf("%d %d", &pos, &val);
				f.update(pos + 1, val);
			}
			else
			{
				int l, r;
				scanf("%d %d", &l, &r);
				printf("%d\n", f.query(l + 1, r + 1));
			}
		}
	}

	return 0;
}

File: fenwick-tree/csharp/1112.cs
using System;
namespace _1112___Curious_Robin_Hood
{
    class Program
    {
        static void Main(string[] args)
        {
            int T = int.Parse(Console.ReadLine());
            string[] data;
            int n, q;

            for (int tc = 1; tc <= T; tc++)
            {
                data = Console.ReadLine().Split();
                n = int.Parse(data[0]);
                q = int.Parse(data[1]);

                Fenwick_Tree ft = new Fenwick_Tree(n);

                data = Console.ReadLine().Split();
                for (int i = 1; i <= n; i++)
                    ft.update(i, int.Parse(data[i - 1]));

                Console.WriteLine("Case {0}:", tc);
                for (int i = 0; i < q; i++)
                {
                    data = Console.ReadLine().Split();
                    if (data[0] == "1")
                    {
                        int pos = int.Parse(data[1]) + 1;
                        int sack = ft.query(pos, pos);
                        ft.update(pos, -sack);
                        Console.WriteLine(sack);
                    }
                    else if (data[0] == "2")
                    {
                        int pos = int.Parse(data[1]) + 1;
                        int val = int.Parse(data[2]);
                        ft.update(pos, val);
                    }
                    else
                    {
                        int l = int.Parse(data[1]) + 1;
                        int r = int.Parse(data[2]) + 1;
                        Console.WriteLine(ft.query(l, r));
                    }
                }
            }
        }
        private class Fenwick_Tree
        {
            private int _n;
            private int[] ft;
            public Fenwick_Tree(int n)
            {
                _n = n;
                ft = new int[_n + 1];
            }
            private int lowbit(int v)
            {
                return v & (-v);
            }
            internal int query(int u, int v)
            {
                return query(v) - query(u - 1);
            }
            private int query(int pos)
            {
                int sum = 0;
                for (; pos > 0; pos -= lowbit(pos))
                    sum += ft[pos];
                return sum;
            }
            internal void update(int pos, int val)
            {
                for (; pos <= _n; pos += lowbit(pos))
                    ft[pos] += val;
            }
        }
    }
}
